Digitizing logger feature

nirvn opened the pull request:
This PR adds a new digitizing logger features which stores each vertex entered while digitizing in QField into point features in a 'digitizing logs layer' (a project setting that's setup via QGIS). Each point feature has access to the snapping result context, to the position context, as well as a couple of digitizing logger variables of its own: @digitizing_type, @digitizing_layer_name, @digitizing_layer_id, @digitizing_datetime.

The @digitizing_type variable stores the kind of action the user was doing when entering vertices (feature addition, ring fill, split, reshape, etc.). The other variables are self explanatory.

To test this out, you can use the QGIS python console, select the point layer in a project you want to use as digitizing logs layer and type the following in the console:

`QgsProject.instance().writeEntry('qfieldsync','digitizingLogsLayer',iface.activeLayer().id())`

This will ultimately be set up in a user friendly manner via the QFieldSync plugin.

Note: digitizing operations that are aborted will not have its vertices logged.

qfield-fairy commented:
🎉 Ta-daaa, freshly created APKs are available for a72d89bb439fe96977b9bfdbccaa4619b9979dc9:
  - [**arm64_v8a**](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1971-77631759b3b1db6540225520c0e31637411467ba-arm64_v8a.apk)

Other architectures: [armv7](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1971-77631759b3b1db6540225520c0e31637411467ba-armv7.apk), [x86_64](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1971-77631759b3b1db6540225520c0e31637411467ba-x86_64.apk), [x86](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1971-77631759b3b1db6540225520c0e31637411467ba-x86.apk)

qfield-fairy commented:
🎉 Ta-daaa, freshly created APKs are available for 91cca1a962ae3822b1d8f609eb7630987df3bab0:
  - [**arm64_v8a**](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1971-a5a4ca791e90c6457876683ee9422f2f60a4d45a-arm64_v8a.apk)

Other architectures: [armv7](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1971-a5a4ca791e90c6457876683ee9422f2f60a4d45a-armv7.apk), [x86_64](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1971-a5a4ca791e90c6457876683ee9422f2f60a4d45a-x86_64.apk), [x86](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1971-a5a4ca791e90c6457876683ee9422f2f60a4d45a-x86.apk)

qfield-fairy commented:
🎉 Ta-daaa, freshly created APKs are available for fec568850e3ff150547a3bd53773ac3e5edec760:
  - [**arm64_v8a**](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1971-bd045fc00f1dc1c18a3ea01e69ba429458b56074-arm64_v8a.apk)

Other architectures: [armv7](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1971-bd045fc00f1dc1c18a3ea01e69ba429458b56074-armv7.apk), [x86_64](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1971-bd045fc00f1dc1c18a3ea01e69ba429458b56074-x86_64.apk), [x86](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1971-bd045fc00f1dc1c18a3ea01e69ba429458b56074-x86.apk)

qfield-fairy commented:
🎉 Ta-daaa, freshly created APKs are available for 79f0bca6adc235b560eb8e300b3e69f37acc7e6f:
  - [**arm64_v8a**](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1971-80c3b7afa075acee6fa4fb47dcdf0b4083035fde-arm64_v8a.apk)

Other architectures: [armv7](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1971-80c3b7afa075acee6fa4fb47dcdf0b4083035fde-armv7.apk), [x86_64](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1971-80c3b7afa075acee6fa4fb47dcdf0b4083035fde-x86_64.apk), [x86](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1971-80c3b7afa075acee6fa4fb47dcdf0b4083035fde-x86.apk)

qfield-fairy commented:
🎉 Ta-daaa, freshly created APKs are available for f7e5755a017ae585de3f670aee52b73c1ba60b19:
  - [**arm64_v8a**](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1971-2841731d643ef908ea86090393b7db39a918a67a-arm64_v8a.apk)

Other architectures: [armv7](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1971-2841731d643ef908ea86090393b7db39a918a67a-armv7.apk), [x86_64](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1971-2841731d643ef908ea86090393b7db39a918a67a-x86_64.apk), [x86](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1971-2841731d643ef908ea86090393b7db39a918a67a-x86.apk)

qfield-fairy commented:
🎉 Ta-daaa, freshly created APKs are available for 01428e20d0b70d0fbc66f915e821709a9d22ea96:
  - [**arm64_v8a**](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1971-3e8c00ac2891ce90dcb2edb87dacbd64b2fdef2c-arm64_v8a.apk)

Other architectures: [armv7](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1971-3e8c00ac2891ce90dcb2edb87dacbd64b2fdef2c-armv7.apk), [x86_64](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1971-3e8c00ac2891ce90dcb2edb87dacbd64b2fdef2c-x86_64.apk), [x86](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1971-3e8c00ac2891ce90dcb2edb87dacbd64b2fdef2c-x86.apk)

qfield-fairy commented:
🎉 Ta-daaa, freshly created APKs are available for e691cde04b6b9fb88ea31d710deb31d5f6a10a23:
  - [**arm64_v8a**](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1971-1aed6e91b8bb3cc0f7c9ce02a20054d6f43dbf09-arm64_v8a.apk)

Other architectures: [armv7](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1971-1aed6e91b8bb3cc0f7c9ce02a20054d6f43dbf09-armv7.apk), [x86_64](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1971-1aed6e91b8bb3cc0f7c9ce02a20054d6f43dbf09-x86_64.apk), [x86](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1971-1aed6e91b8bb3cc0f7c9ce02a20054d6f43dbf09-x86.apk)

qfield-fairy commented:
🎉 Ta-daaa, freshly created APKs are available for 78ff3f0e8f46d138113ce54674180e6d8bb2d923:
  - [**arm64_v8a**](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1971-7ad38080bde0fea590ff23d53bab8001766a7052-arm64_v8a.apk)

Other architectures: [armv7](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1971-7ad38080bde0fea590ff23d53bab8001766a7052-armv7.apk), [x86_64](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1971-7ad38080bde0fea590ff23d53bab8001766a7052-x86_64.apk), [x86](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1971-7ad38080bde0fea590ff23d53bab8001766a7052-x86.apk)

geckeau commented:
Nice progress! After some quick testing I get the following results.

- When adding a line and a polygon, the three position-based vertices are correctly logged.
- But the position-based line vertex is logged twice.
- From the line feature, only the last manually added vertex is logged, the other manual vertices are not logged.
- From the polygon feature, no manually (with the stylus pen) added vertex is correctly logged. (Neither from inserting pointwise with the stylus pen, nor from freehand digitizing with the stylus pen)
- An additional point appears on the logging layer that lies outside of the polygon. (I suppose, this was the position of the cursor when i confirmed the polygon with the green tick. But actually no vertex was set at this position)

![QFieldExtensionTest](https://user-images.githubusercontent.com/85615456/121333406-147c7c00-c919-11eb-925a-e84bb53b3a2c.png)

nirvn commented:
@geckeau , thanks for the feedback, looking into it.

nirvn commented:
@geckeau , stylus and freehand should work now. I've not been able to replicate the random point being logged. If you can provide steps to reproduce while you're trying the new APKs coming in a couple of minutes, that'd be great.

qfield-fairy commented:
🎉 Ta-daaa, freshly created APKs are available for 988a788e3ba308c9628f1252c5375c117c555bd5:
  - [**arm64_v8a**](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1971-8d95b64d01fa2dddb0a9cf6327a8c9c254044474-arm64_v8a.apk)

Other architectures: [armv7](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1971-8d95b64d01fa2dddb0a9cf6327a8c9c254044474-armv7.apk), [x86_64](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1971-8d95b64d01fa2dddb0a9cf6327a8c9c254044474-x86_64.apk), [x86](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1971-8d95b64d01fa2dddb0a9cf6327a8c9c254044474-x86.apk)

geckeau commented:
@nirvn Thanks for your answers. Unfortunately with the Demo project that I was provided by Matthias Kuhn and the new APK, logging doesn't work at all any more. Do I also have to update the demo project file somehow?

m-kuhn commented:
Updated project file sent by mail

nirvn commented:
@geckeau , is the updated project working out OK?

m-kuhn commented:
@qfield-fairy style please

qfield-fairy commented:
🎉 Ta-daaa, freshly created APKs are available for 71a02221cd2d8197b6a38dae7e3ed1429bdecf71:
  - [**arm64_v8a**](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1971-325c62f9d7f94b88468b0f1df73b7ec9f70bae42-arm64_v8a.apk)

Other architectures: [armv7](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1971-325c62f9d7f94b88468b0f1df73b7ec9f70bae42-armv7.apk), [x86_64](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1971-325c62f9d7f94b88468b0f1df73b7ec9f70bae42-x86_64.apk), [x86](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1971-325c62f9d7f94b88468b0f1df73b7ec9f70bae42-x86.apk)

qfield-fairy commented:
🎉 Ta-daaa, freshly created APKs are available for 858326b524d96eddf6390ea961caf2ad633d633c:
  - [**arm64_v8a**](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1971-3df1ada8b1a9daa44e2225e2439c5924ea0f2763-arm64_v8a.apk)

Other architectures: [armv7](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1971-3df1ada8b1a9daa44e2225e2439c5924ea0f2763-armv7.apk), [x86_64](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1971-3df1ada8b1a9daa44e2225e2439c5924ea0f2763-x86_64.apk), [x86](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1971-3df1ada8b1a9daa44e2225e2439c5924ea0f2763-x86.apk)

geckeau commented:
@m-kuhn Thanks for the updated project file
@nirvn  Thanks for the work. Here come the next test results:

Not able to replicate the randomly logged vertex. Pen vertices are now logged. Current bugs:
- line vertices are logged twice (see image)
- No non-position based vertices (touch/pen) added for reshape tool and fill ring tool
- No vertices added  when editing a feature with the vertex tool
- Log vertices are added even when feature is aborted when assigning the attribute values, so before final confirmation of the feature

General observation/bug
- split tool assigns original attribute values only to one of the two split features (different behavior Beta version Vs. QField 1.9.6)

![image](https://user-images.githubusercontent.com/85615456/121531562-d0fc3d80-c9fe-11eb-818f-9c842646b793.png)

nirvn commented:
@geckeau , fantastic report.

Questions:
- only line vertices are added twice? Ie not a problem with polygons?
- what do you mean by no non-position added? If you can expand your description it'd help me. In English ou en francais, ca me va :)
- you are correct, ATM we add vertices after a confirmed digitizing action, which happens before someone ultimately safe or discard a feature. If that's a big issue for your workflow we can try and adjust this behavior

On the split feature, can you confirm it's also a different behavior in QGIS 3.19 (what will become 3.20)? I noticed this during the development of this feature, tested against QGIS 3.19 and noticed the same behavior.

Cheers.

nirvn commented:
@geckeau , regarding vertices added twice for lines are all vertices added twice or only the last one? I assume it's when you add a new feature, am I correct?

geckeau commented:
I notice that actually I still get random vertex log points when using the tools fillring, reshape and split. All appear in places where I never knowingly set a vertex

![image](https://user-images.githubusercontent.com/85615456/121539402-f0e32f80-ca05-11eb-8484-1258924a5b77.png)

nirvn commented:
@geckeau , do you confirm your digitizing via hitting the check icon confirm button or by long pressing the pen on the screen?

geckeau commented:
>     * only line vertices are added twice? Ie not a problem with polygons?
It seems like line vertices are always logged twice. Polygon vertices occassionally. I do not know why or when.

>     * what do you mean by no non-position added? If you can expand your description it'd help me. In English ou en francais, ca me va :)

vertices that i add manually with the pen or touch, so without snapping the cursor to my current position

>     * you are correct, ATM we add vertices after a confirmed digitizing action, which happens before someone ultimately safe or discard a feature. If that's a big issue for your workflow we can try and adjust this behavior

It would be very helpful to avoid this and only add vertices for features that are actually saved. Thanks in advance.


> On the split feature, can you confirm it's also a different behavior in QGIS 3.19 (what will become 3.20)? I noticed this during the development of this feature, tested against QGIS 3.19 and noticed the same behavior.

I do not have QGIS 3.19 installed unfortunately.

> @geckeau , regarding vertices added twice for lines are all vertices added twice or only the last one? I assume it's when you add a new feature, am I correct?

I only really tested lines for adding atm, not for editing. It seems to me that all line vertices are added twice.

I am sorry but I am very restricted on time and not much available for further intensive testing. But I think we described the expected behavior of the QField extension well in advance, so it should be possible for you to do the testing on this basis. I might be able to do some more testing tomorrow afternoon. Thank you for your understanding.

geckeau commented:
> 
> 
> @geckeau , do you confirm your digitizing via hitting the check icon confirm button or by long pressing the pen on the screen?

Until now I always confirmed with the check icon. Did not know about the possibility to long press the pen.

qfield-fairy commented:
🎉 Ta-daaa, freshly created APKs are available for 27284acdfcaac2ed421847147044e1d0bffbc28c:
  - [**arm64_v8a**](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1971-46a4d15fb37c9653068140e47b6797e633b4df41-arm64_v8a.apk)

Other architectures: [armv7](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1971-46a4d15fb37c9653068140e47b6797e633b4df41-armv7.apk), [x86_64](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1971-46a4d15fb37c9653068140e47b6797e633b4df41-x86_64.apk), [x86](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1971-46a4d15fb37c9653068140e47b6797e633b4df41-x86.apk)

qfield-fairy commented:
🎉 Ta-daaa, freshly created APKs are available for df9efc2fc2464ba68b7b772c0304c91c560afab6:
  - [**arm64_v8a**](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1971-1504879f09713063b5cf3d3ec438c117c4f3544c-arm64_v8a.apk)

Other architectures: [armv7](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1971-1504879f09713063b5cf3d3ec438c117c4f3544c-armv7.apk), [x86_64](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1971-1504879f09713063b5cf3d3ec438c117c4f3544c-x86_64.apk), [x86](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1971-1504879f09713063b5cf3d3ec438c117c4f3544c-x86.apk)

qfield-fairy commented:
🎉 Ta-daaa, freshly created APKs are available for c1ce6db5ee01381d59bf0563af2ae10cc500d83d:
  - [**arm64_v8a**](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1971-cca4137ddad117e04061f5d2df939160c1e77992-arm64_v8a.apk)

Other architectures: [armv7](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1971-cca4137ddad117e04061f5d2df939160c1e77992-armv7.apk), [x86_64](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1971-cca4137ddad117e04061f5d2df939160c1e77992-x86_64.apk), [x86](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1971-cca4137ddad117e04061f5d2df939160c1e77992-x86.apk)

m-kuhn commented on src/qml/DigitizingToolbar.qml:
```diff
@@ -219,4 +237,10 @@ VisibilityFadingRow {
     mapSettings.setCenter( rubberbandModel.currentCoordinate )
   }
+
+  function confirming()
```
What's the diff between `confirming()` and `confirm()`?

nirvn replied:
The function confirming() triggers code to finalize the digitizing, while signal confirm() is what's emitted when confirming is done. Not sure if that's helpful :) I hesitated between that and function confirm() signal confirmed().

m-kuhn replied:
I think the second one would be clearer in terms of semantic api

qfield-fairy commented:
🎉 Ta-daaa, freshly created APKs are available for 5bd75a7221b6337f5912b05f94dbe28b86d22054:
  - [**arm64_v8a**](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1971-5fdc5c41c4159d89dac6d52f9aaf5fda0506f79e-arm64_v8a.apk)

Other architectures: [armv7](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1971-5fdc5c41c4159d89dac6d52f9aaf5fda0506f79e-armv7.apk), [x86_64](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1971-5fdc5c41c4159d89dac6d52f9aaf5fda0506f79e-x86_64.apk), [x86](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1971-5fdc5c41c4159d89dac6d52f9aaf5fda0506f79e-x86.apk)

qfield-fairy commented:
🎉 Ta-daaa, freshly created APKs are available for 0a80488f6a1174ceed788a1da4222a22606012dd:
  - [**arm64_v8a**](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1971-bb18db12b4eef1741ce3954e86dab4150869806f-arm64_v8a.apk)

Other architectures: [armv7](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1971-bb18db12b4eef1741ce3954e86dab4150869806f-armv7.apk), [x86_64](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1971-bb18db12b4eef1741ce3954e86dab4150869806f-x86_64.apk), [x86](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1971-bb18db12b4eef1741ce3954e86dab4150869806f-x86.apk)

qfield-fairy commented:
🎉 Ta-daaa, freshly created APKs are available for 20d06cb822006d6bca6276f6ab4e5a8c667d866e:
  - [**arm64_v8a**](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1971-a28b3e863284d5a27b3deff3aa29b2a2db677cfb-arm64_v8a.apk)

Other architectures: [armv7](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1971-a28b3e863284d5a27b3deff3aa29b2a2db677cfb-armv7.apk), [x86_64](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1971-a28b3e863284d5a27b3deff3aa29b2a2db677cfb-x86_64.apk), [x86](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1971-a28b3e863284d5a27b3deff3aa29b2a2db677cfb-x86.apk)

nirvn commented:
@m-kuhn , confirm naming adjusted.

qfield-fairy commented:
🎉 Ta-daaa, freshly created APKs are available for 36c6282a5ee58c93ca23eb3a34e61d2e609dc781:
  - [**arm64_v8a**](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1971-85044746dd676fbcc683b1f9542e3b7f8700ca10-arm64_v8a.apk)

Other architectures: [armv7](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1971-85044746dd676fbcc683b1f9542e3b7f8700ca10-armv7.apk), [x86_64](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1971-85044746dd676fbcc683b1f9542e3b7f8700ca10-x86_64.apk), [x86](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1971-85044746dd676fbcc683b1f9542e3b7f8700ca10-x86.apk)

qfield-fairy commented:
🎉 Ta-daaa, freshly created APKs are available for ab8aa3c3071097c8af70c8c7939ce1a704045479:
  - [**arm64_v8a**](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1971-e26fb37e439c53e6c0f46e30a084111a94b79b13-arm64_v8a.apk)

Other architectures: [armv7](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1971-e26fb37e439c53e6c0f46e30a084111a94b79b13-armv7.apk), [x86_64](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1971-e26fb37e439c53e6c0f46e30a084111a94b79b13-x86_64.apk), [x86](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1971-e26fb37e439c53e6c0f46e30a084111a94b79b13-x86.apk)

geckeau commented:
I just got the chance for another round of testing. It's already looking really nice! Very cool, that you successfully implemented to wait with logging until ultimate confirmation or discard. Thanks! :)

Some things are not yet running smooth. I made the following observations:
- a) No vertices added when editing a feature with the vertex tool when having cursor snapped to position (for both lines and polygons)
- b) The reshape tool does not work at all for polygons in the current apk-build. Logging vertices are added but the feature geometry remains unchanged.
- c) Sometimes, after confirming the last vertex, instead of showing the attribute form, the geometry is discarded. But the vertices are logged on the logging layer. Very strange behavior. It happened 3 times, both for polygons and lines, in seemingly random situations. Maybe a little QField crash without really crashing. I can unfortunately not provide a procedure to reproduce atm. Also, one time the attribute form was shown with area = 0 and final confirmation not possible.
- In the attribute tables of the polygon layers, an empty geometry with null values for each attribute was added. I suppose it stems either from b) or from c)
- d) Two random vertices were created at some point with NULL values for all @digitizing_* variables. I'm unfortunately not able to tell when it happened

Non log-layer specific
- Split lines appends '(' before and '_1)' behind string attribute names. So the categorized or rule-based symbology does not work as expected.

geckeau commented:
Also
- The variable @digitizing_layer_name only works, when adding new features, not when editing
- Idem for @digitizing_layer_id.
- When adding a new feature, @digitizing_layer_id is always =0. (I chose data type int. Is this a bad choice? How to set it up correctly?)

nirvn commented:
@geckeau , OK, I think we're getting very close to finishing this feature. I've fixed the problem with geometry editing tools (reshape, split, ring) missing digitizing layer. On that note, @digitizing_layer_id is a string, not an int value. It represents the unique ID of a given layer in your project (it's a permanent and unique string that allows users to reach the right layer even if he/she renames the layer or has duplicate layer names).

I've noticed the reshape is intermittently failing here when using freehand digitizing. I believe it's a separate issue not linked to this feature. I'll look into it but IMHO it shouldn't block this PR. Same for the the split line / polygon issue, which might be a QGIS 3.20 regression or a project setting that doesn't behave the same way now. 

The last digitizing logger specific issue you've raised is the vertex editor's snapping to location issue. I shall investigate now :) 

Thanks for the testing during the making of this feature.

qfield-fairy commented:
🎉 Ta-daaa, freshly created APKs are available for 362a2a35f8d349588185e3d1034eb8ba800094d6:
  - [**arm64_v8a**](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1971-b1fd2412dd931c18e8d317a15f49c283c1f45002-arm64_v8a.apk)

Other architectures: [armv7](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1971-b1fd2412dd931c18e8d317a15f49c283c1f45002-armv7.apk), [x86_64](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1971-b1fd2412dd931c18e8d317a15f49c283c1f45002-x86_64.apk), [x86](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1971-b1fd2412dd931c18e8d317a15f49c283c1f45002-x86.apk)

nirvn commented:
@geckeau , alright, the vertex editor should log modified vertices for all scenarios.

qfield-fairy commented:
🎉 Ta-daaa, freshly created APKs are available for 80758ec51bf8fe7cd185509387fc8fd7181e5885:
  - [**arm64_v8a**](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1971-4d8b78818055edd5f1ca110fec36f740c2431695-arm64_v8a.apk)

Other architectures: [armv7](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1971-4d8b78818055edd5f1ca110fec36f740c2431695-armv7.apk), [x86_64](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1971-4d8b78818055edd5f1ca110fec36f740c2431695-x86_64.apk), [x86](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1971-4d8b78818055edd5f1ca110fec36f740c2431695-x86.apk)

qfield-fairy commented:
🎉 Ta-daaa, freshly created APKs are available for 10138fdcd135fee9d9fae47c38005ce74bf8e7bb:
  - [**arm64_v8a**](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1971-9d1aa39ed0dfa9555f1f3e7c84b02aae89db25a5-arm64_v8a.apk)

Other architectures: [armv7](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1971-9d1aa39ed0dfa9555f1f3e7c84b02aae89db25a5-armv7.apk), [x86_64](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1971-9d1aa39ed0dfa9555f1f3e7c84b02aae89db25a5-x86_64.apk), [x86](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1971-9d1aa39ed0dfa9555f1f3e7c84b02aae89db25a5-x86.apk)

nirvn commented:
@geckeau , I've looked into the split issue, and happy to report QField is behaving as it should :wink: 

In your project, the 'Struktur' field has a [x] unique constraint checked, which results in the newly created feature created when splitting the original polygon having a _X appended to the end of its value in order to insure the uniqueness of the field value.

If you uncheck the [x] unique constraint, you'll see geometries being split just fine in QField.

qfield-fairy commented:
🎉 Ta-daaa, freshly created APKs are available for c0898bee883b0ca0d78a84d5e596384bfbc4a1e1:
  - [**arm64_v8a**](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1971-4d35714cc6dbb20cdc8fa722ce68ee618b2cc2c2-arm64_v8a.apk)

Other architectures: [armv7](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1971-4d35714cc6dbb20cdc8fa722ce68ee618b2cc2c2-armv7.apk), [x86_64](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1971-4d35714cc6dbb20cdc8fa722ce68ee618b2cc2c2-x86_64.apk), [x86](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1971-4d35714cc6dbb20cdc8fa722ce68ee618b2cc2c2-x86.apk)

m-kuhn commented on src/core/digitizinglogger.cpp:
```diff
@@ -0,0 +1,235 @@
+
+void DigitizingLogger::setPositionInformation( const GnssPositionInformation &positionInformation )
+{
+  mPositionInformation = positionInformation;
```
Is there no recursion guard here on purpose?

m-kuhn commented on src/core/digitizinglogger.cpp:
```diff
@@ -0,0 +1,235 @@
+
+void DigitizingLogger::setTopSnappingResult( const SnappingResult &topSnappingResult )
+{
+  mTopSnappingResult = topSnappingResult;
```
Recursion guard?

m-kuhn commented on src/core/digitizinglogger.cpp:
```diff
@@ -0,0 +1,235 @@
+
+  if ( mProject )
+  {
+    disconnect( mProject, &QgsProject::readProject, this, &DigitizingLogger::findLogsLayer );
```
If that happens (which it won't ;) ) don't we also need to clear layers etc?

nirvn replied:
@m-kuhn , not the digitizing layer, and the logs layer gets reset in findLogsLayer() function which is called in setProject(...)

m-kuhn commented on src/core/digitizinglogger.cpp:
```diff
@@ -0,0 +1,235 @@
+    const QString logsLayerId = mProject->readEntry( QStringLiteral( "qfieldsync" ), QStringLiteral( "digitizingLogsLayer" ) );
+    if ( !logsLayerId.isEmpty() )
+    {
+      QgsLayerTreeLayer *item = mProject->layerTreeRoot()->findLayer( logsLayerId );
```
More of a qf sync question: do we always have the offline id here and never the online id?

m-kuhn reviewed (approved):
Looks pretty awesome to me

nirvn commented:
@geckeau , alright, I've  fixed a reshape tool issue which you likely were hitting while stress testing your project (https://github.com/opengisch/QField/pull/1996) -- it's also likely you ended up with "invisible" polygons after splitting polygons and ending up with unique values for the Struktur field the project uses to style the layer. Those are treacherous, you don't see them but they still act as feature that aren't allowed to be overlapped. You'd do a reshape operation and end up overlapping a split feature not knowing it is there to begin with.

nirvn commented:
@m-kuhn , review addressed.

nirvn commented:
@qfield-fairy style please

qfield-fairy commented:
🎉 Ta-daaa, freshly created APKs are available for c212fd96344e26030d97d235de14e9235d4236f1:
  - [**arm64_v8a**](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1971-51a7d79caaa5ddbda6e9bb4c4585712398462da6-arm64_v8a.apk)

Other architectures: [armv7](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1971-51a7d79caaa5ddbda6e9bb4c4585712398462da6-armv7.apk), [x86_64](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1971-51a7d79caaa5ddbda6e9bb4c4585712398462da6-x86_64.apk), [x86](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1971-51a7d79caaa5ddbda6e9bb4c4585712398462da6-x86.apk)

qfield-fairy commented:
🎉 Ta-daaa, freshly created APKs are available for fa9862ce9333edbf285ff2965ace1ab61cc73d91:
  - [**arm64_v8a**](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1971-6ee752bbd340c955e619b0c5f9d69bb5360906c4-arm64_v8a.apk)

Other architectures: [armv7](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1971-6ee752bbd340c955e619b0c5f9d69bb5360906c4-armv7.apk), [x86_64](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1971-6ee752bbd340c955e619b0c5f9d69bb5360906c4-x86_64.apk), [x86](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1971-6ee752bbd340c955e619b0c5f9d69bb5360906c4-x86.apk)

geckeau commented:
> @geckeau , alright, I've fixed a reshape tool issue which you likely were hitting while stress testing your project (#1996) -- it's also likely you ended up with "invisible" polygons after splitting polygons and ending up with unique values for the Struktur field the project uses to style the layer.
> 
> Those invisible polygons are treacherous, you don't see them but they still act as feature that aren't allowed to be overlapped. You'd do a reshape operation and end up overlapping a split feature not knowing it is there to begin with. I suspect this is why you ended up with empty geometries when adding new features too.

@nirvn Thanks for all the fixes! The tool looks great now. Unfortunately the reshaping still makes trouble. For my project as well as for a very basic demo project with only a Polygon and Multipolygonlayer. I will provide a description in pull [https://github.com/opengisch/QField/pull/1996](url)

Could you please have a look at my mail regarding the QFieldSync .zip? Thank you!
Cheers

Fertuesta25 commented:
I have tried to use this function but no success, any suggestions?
